### PR TITLE
support for manually specified dependencies

### PR DESCRIPTION
**Pull request by zlav**
# Overview
[Link to the PRD](https://github.com/fossas/team-analysis/issues/425)

The goal of this PR is to give users the ability to manually add dependencies for their projects. This allows users with non standard package managers to still leverage the FOSSA ecosystem without waiting on us to create first-class support. This also means we no longer have to debate whether support is worth adding or if adding support will hurt UX down the line. This PR allows us to create custom scripts to support customers for any ecosystem.

## Acceptance criteria

This PR should introduce an easy-to-use way for customers to upload dependencies to FOSSA.
- `fossa-deps.yaml` parses correctly.
- Parsing fails if the file is misformatted, fields are missing, or unsupported values are provided.
- The dependencies are merged with normal `fossa analyze` results and uploaded alongside them.

## Testing plan

Validate that the `fossa-deps.yaml` file works as intended.
- Scan a normal project with spectrometer that has >1 dependency.
- Construct a custom `fossa-deps.yaml` file and run `fossa analyze`.
- Validate that the manually added deps are included.
Tests include
- One for a simple `fossa-deps.yaml` file that checks for optional fields.
- One test for a poorly formatted file that should fail.

## Risks

The largest risk with this PR is accidentally committing to behavior that we regret in 6 months. Outside of that, there is almost 0 chance of regression.

## References

Closes https://github.com/fossas/team-analysis/issues/425

## Checklist

- [X] I added tests for this PR's change.
- [X] I linted (`hlint`) and formatted (`ormolu`) any files I touched in this PR.
- [X] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [X] I updated `CHANGELOG.md`. If this PR does not mark a release update the `#Unreleased` section.
- [X] I linked this PR to any referenced GitHub issues.

### CODE REVIEW

File: src/Strategy/UserSpecified/YamlDependencies.hs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+import Graphing (Graphing)
+import qualified Graphing
+import Path
+import Types
+
+discover :: (Has ReadFS sig m, Has Diagnostics sig m, Has ReadFS rsig run, Has Diagnostics rsig run) => Path Abs Dir -> m [DiscoveredProject run]
+discover dir = map mkProject <$> findProjects dir
+
+-- Only search for fossa-deps.yaml in the root directory. We can extend this to subdirectories in the future.
+findProjects :: (Has ReadFS sig m, Has Diagnostics sig m) => Path Abs Dir -> m [UserDependenciesYamlProject]
```
Comment (cnr) [inline src/Strategy/UserSpecified/YamlDependencies.hs:30]: Given that we only want to find the file in the root, we can skip the walk and simplify this to something like:
```hs
findProjects dir = do
  exists <- doesFileExist $(mkRelFile ".fossa.yml")
  if exists
    then .. -- make the project
    else pure []
```

File: src/Strategy/UserSpecified/YamlDependencies.hs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+findProjects = walk' $ \dir _ files -> do
+  _ <- traceM $ show files
+  case findFileNamed "fossa-deps.yaml" files of
```
Comment (cnr) [inline src/Strategy/UserSpecified/YamlDependencies.hs:33]: I think `.yml` is much more common as an extension for yaml files; we should probably use it instead.

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -26,8 +26,8 @@ _Example:_ Closes https://github.com/example/repo/issues/123.
...
 
-- [ ] I added tests for this PR's change.
-- [ ] I linted (`hlint`) and formatted (`ormolu`) any files I touched in this PR.
-- [ ] If this PR introduced a user-visible change, I added documentation into `docs/`.
-- [ ] If this PR marked a release, I updated `CHANGELOG.md`.
-- [ ] I linked this PR to any referenced GitHub issues.
+- [] I added tests for this PR's change.
+- [] I linted (`hlint`) and formatted (`ormolu`) any files I touched in this PR.
+- [] If this PR introduced a user-visible change, I added documentation into `docs/`.
+- [] I updated `CHANGELOG.md`. If the PR did not mark a release, update the `#Unreleased section at the top`.
```
Comment (cnr) [inline .github/pull_request_template.md:32]:
```suggestion
- [] I updated the `Unreleased` section in `CHANGELOG.md` with a description of my changes
```

File: docs/userguide.md [hunk 4]
```diff
@@ -169,6 +170,40 @@ We support the following archive formats:
 - `.jar`
 - `.rpm`
 
+### Manually specifying dependencies
+
+FOSSA offers a way to manually upload dependencies provided we support the dependency type. Manually specifying dependencies is very helpful in the event your package manager is unsupported or you are using a custom and nonstandard dependency management solution.
+
+The FOSSA CLI will automatically read `fossa-deps.yaml` in the root directory when `fossa analyze` is run and parse dependencies from it.
+
+> Note: You can use a script to reconstruct this file before `fossa analyze` is run, which would allow you to create your own custom strategy.
```
Comment (cnr) [inline docs/userguide.md:179]: Unsure about the wording here; maybe something like:

```suggestion
> Tip: Use a script to generate this file before running `fossa analyze`
```

File: docs/userguide.md [hunk 5]
```diff
@@ -169,6 +170,40 @@ We support the following archive formats:
...
+- `hackage` - Haskell dependencies found at [Hackage](https://hackage.haskell.org/).
+- `hex` - Erlang and Elixir dependencies that are found at [Hex.pm](https://hex.pm/).
+- `maven` - Maven dependencies that can be found at many different sources. Specified as `package: javax.xml.bind:jaxb-api` where the convention is `groupId:artifactId`.
+- `npm` - Javascript dependencies found at [npmjs.com](https://www.npmjs.com/).
+- `nuget` - .NET dependencies found at [NuGet.org](https://www.nuget.org/).
+- `python` - Python dependencies found at [Pypi.org](https://pypi.org/).
+- `cocoapods` - Swift and Objective-C dependencies found at [Cocoapods.org](https://cocoapods.org/).
+- `url` - The URL type allows you to specify only the download location of a compressed file in the `package` field and the FOSSA backend will attempt to download and scan it. Example for a Maven dependency `https://repo1.maven.org/maven2/aero/m-click/mcpdf/0.2.3/mcpdf-0.2.3-jar-with-dependencies.jar$`. The `version` field will be ignored for `url` type dependencies.
```
Comment (cnr) [inline docs/userguide.md:205]: extraneous dollar sign

```suggestion
- `url` - The URL type allows you to specify only the download location of a compressed file in the `package` field and the FOSSA backend will attempt to download and scan it. Example for a Maven dependency `https://repo1.maven.org/maven2/aero/m-click/mcpdf/0.2.3/mcpdf-0.2.3-jar-with-dependencies.jar`. The `version` field will be ignored for `url` type dependencies.
```

**Review by cnr — APPROVED** [review]
Looks great to me; mostly nits

File: src/Strategy/UserSpecified/YamlDependencies.hs (hunk 1 above)
Comment (zlav) [reply src/Strategy/UserSpecified/YamlDependencies.hs:30]: For some reason, I kept thinking that `mkRelFile` was actually creating a file at that location, thanks.

**Pull request merged**